src/package-request.js

```diff
@@ -193,6 +193,7 @@ export default class PackageRequest {
     invariant(ref, 'Resolved package info has no package reference');
     ref.addRequest(this);
     ref.addPattern(this.pattern, resolved);
+    ref.addOptional(this.optional);
```

Comment:
The new test case revealed another bug here: if a package is listed as optional by one package and as a strict dependency by another, it should be non-optional at the end. However, due to incorrect request caching below and the lack of this line here, it's optional status was random :(

Reply:
nice catch!